src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Legacy/LegacySyntaxNodeExtensions.cs

```diff
@@ -258,49 +318,75 @@ public static IEnumerable<SyntaxNode> FlattenSpans(this SyntaxNode node)
         }
     }
 
-    public static SyntaxNode PreviousSpan(this SyntaxNode node)
+    public static SyntaxNode? PreviousSpan(this SyntaxNode node)
     {
-        if (node == null)
+        if (node is null)
         {
             throw new ArgumentNullException(nameof(node));
         }
 
+        var spanData = s_spanDataTable.GetOrCreateValue(node);
+
+        if (spanData.PreviousComputed)
```

Comment:
I'm also not certain that the read of `spanData.Previous` can't be reordered before `spanData.PreviousComputed` by speculative execution, as there is no real data dependence between the two.

Reply:
To simplify, I ended up just locking on `spanData`.